// chainlink/dist/contracts/ChainlinkedFactory.ts
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */

import { Contract, ContractFactory, Signer } from "ethers";
import { Provider } from "ethers/providers";
import { UnsignedTransaction } from "ethers/utils/transaction";

import { Chainlinked } from "./Chainlinked";

export class ChainlinkedFactory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }
  deploy(): Promise<Chainlinked> {
    return super.deploy() as Promise<Chainlinked>;
  }
  getDeployTransaction(): UnsignedTransaction {
    return super.getDeployTransaction();
  }
  attach(address: string): Chainlinked {
    return super.attach(address) as Chainlinked;
  }
  connect(signer: Signer): ChainlinkedFactory {
    return super.connect(signer) as ChainlinkedFactory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Chainlinked {
    return new Contract(address, _abi, signerOrProvider) as Chainlinked;
  }
}

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkRequested",
    type: "event"
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkFulfilled",
    type: "event"
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkCancelled",
    type: "event"
  }
];

const _bytecode =
  "0x60806040526001600455348015601457600080fd5b5060358060226000396000f3006080604052600080fd00a165627a7a723058208b59bbef4d9a55977be46d1446f221a2fdebd0efecd17d2df64cf2d48e5f4fe60029";
