// chainlink/dist/contracts/MaliciousChainlinkedFactory.ts
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */

import { Contract, ContractFactory, Signer } from "ethers";
import { Provider } from "ethers/providers";
import { UnsignedTransaction } from "ethers/utils/transaction";

import { MaliciousChainlinked } from "./MaliciousChainlinked";

export class MaliciousChainlinkedFactory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }
  deploy(): Promise<MaliciousChainlinked> {
    return super.deploy() as Promise<MaliciousChainlinked>;
  }
  getDeployTransaction(): UnsignedTransaction {
    return super.getDeployTransaction();
  }
  attach(address: string): MaliciousChainlinked {
    return super.attach(address) as MaliciousChainlinked;
  }
  connect(signer: Signer): MaliciousChainlinkedFactory {
    return super.connect(signer) as MaliciousChainlinkedFactory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MaliciousChainlinked {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as MaliciousChainlinked;
  }
}

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkRequested",
    type: "event"
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkFulfilled",
    type: "event"
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32"
      }
    ],
    name: "ChainlinkCancelled",
    type: "event"
  }
];

const _bytecode =
  "0x608060405260016004556001600655348015601957600080fd5b5060358060276000396000f3006080604052600080fd00a165627a7a72305820be8454cbafe1663b485a44375bc60aec19326c368f999f8f2b6530dd958185660029";
